graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/NodeMetadataSocialProofGenerator.java

```diff
@@ -0,0 +1,183 @@
+      // We choose ArrayMap over OpenHashMap since the request will have at most a few edge types,
+      // usually less than five.
+      socialProofs.put(metadataId, new Byte2ObjectArrayMap<>());
+      socialProofWeights.put(metadataId, 0);
+    }
+    Byte2ObjectMap<Long2ObjectMap<LongSet>> socialProofMap = socialProofs.get(metadataId);
+
+    // Get the user to tweets map variable by the engagement type.
+    if (!socialProofMap.containsKey(edgeType)) {
+      // We choose to use an OpenHashMap since a single edge type may have dozens or even hundreds
+      // of user seed ids associated.
+      socialProofMap.put(edgeType, new Long2ObjectOpenHashMap<>());
+    }
+    Long2ObjectMap<LongSet> userToTweetsMap = socialProofMap.get(edgeType);
+
+    // Add the connecting user to the user map.
+    if (!userToTweetsMap.containsKey(leftNode)) {
+      // We choose to use an OpenHashSet since a single user may engage with dozens or even
+      // hundreds of tweet ids.
+      userToTweetsMap.put(leftNode, new LongOpenHashSet());
+    }
+    LongSet connectingTweets = userToTweetsMap.get(leftNode);
+
+    // Add the connecting tweet to the tweet set.
+    if (!connectingTweets.contains(rightNode)) {
+      connectingTweets.add(rightNode);
+    }
+  }
+
+  /**
+   * Collect social proofs for a given {@link SocialProofRequest}.
+   *
+   * @param request contains a set of input metadata ids and a set of seed users.
+   */
+  private void collectRecommendations(NodeMetadataSocialProofRequest request) {
+    socialProofs.clear();
+    socialProofWeights.clear();
+    IntSet inputNodeMetadataIds = request.getNodeMetadataIds();
+    ByteSet socialProofTypes = new ByteArraySet(request.getSocialProofTypes());
+    byte nodeMetadataType = (byte) this.recommendationType.getValue();
+
+    // Iterate through the set of seed users with weights.
+    for (Long2DoubleMap.Entry entry: request.getLeftSeedNodesWithWeight().long2DoubleEntrySet()) {
+      long leftNode = entry.getLongKey();
+      double weight = entry.getDoubleValue();
+      NodeMetadataMultiSegmentIterator edgeIterator =
+        (NodeMetadataMultiSegmentIterator) graph.getLeftNodeEdges(leftNode);
+      if (edgeIterator == null) continue;
+
+      // For each seed user node, we traverse all of its outgoing edges, up to the MAX_EDGES_PER_NODE.
+      int numEdgePerNode = 0;
+      while (edgeIterator.hasNext() && numEdgePerNode++ < MAX_EDGES_PER_NODE) {
+        long rightNode = idMask.restore(edgeIterator.nextLong());
+        byte edgeType = edgeIterator.currentEdgeType();
+        if (!socialProofTypes.contains(edgeType)) continue;
+
+        IntArrayIterator metadataIterator =
+          (IntArrayIterator) edgeIterator.getRightNodeMetadata(nodeMetadataType);
+        if (metadataIterator == null) continue;
+
+        // For each of the accepted edges, we traverse the metadata ids for the specified node
+        // metadata type (either RecommendationType.HASHTAG or RecommendationType.URL).
```

Comment:
sometimes clients might want to fetch the social proof for both hashtags and urls in one call, can we support that?